From 4cc6af78f4f622ff490cbeb064adf23a20f97c3e Mon Sep 17 00:00:00 2001
From: stonar96 <minecraft.stonar96@gmail.com>
Date: Tue, 3 Oct 2017 17:39:19 +0200
Subject: [PATCH] Valid login

---
 .../net/minecraft/src/NetLoginHandler.java                 | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/sources/minecraft_server/net/minecraft/src/NetLoginHandler.java b/sources/minecraft_server/net/minecraft/src/NetLoginHandler.java
index 40a9980..55f2768 100644
--- a/sources/minecraft_server/net/minecraft/src/NetLoginHandler.java
+++ b/sources/minecraft_server/net/minecraft/src/NetLoginHandler.java
@@ -80,6 +80,20 @@ public class NetLoginHandler extends NetHandler
             }
             return;
         }
+        if(!username.matches("[A-Za-z0-9_]{2,16}"))//Mod: Valid login
+        {//Mod: Valid login
+            kickUser("Invalid username, try: [A-Za-z0-9_]{2,16}");//Mod: Valid login
+            return;//Mod: Valid login
+        }//Mod: Valid login
+        for(int i = 0; i < mcServer.configManager.playerEntities.size(); i++)//Mod: Valid login
+        {//Mod: Valid login
+            EntityPlayerMP entityplayermp = (EntityPlayerMP)mcServer.configManager.playerEntities.get(i);//Mod: Valid login
+            if(entityplayermp.username.equalsIgnoreCase(username))//Mod: Valid login
+            {//Mod: Valid login
+                kickUser("You are already online");//Mod: Valid login
+                return;//Mod: Valid login
+            }//Mod: Valid login
+        }//Mod: Valid login
         if(!mcServer.onlineMode)
         {
             doLogin(packet1login);
