From 4716fe21735a33f3e2336e771352aa2878477fa2 Mon Sep 17 00:00:00 2001
From: stonar96 <minecraft.stonar96@gmail.com>
Date: Sun, 9 Sep 2018 03:39:59 +0200
Subject: [PATCH] Add basic event system


diff --git a/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/CancelablePlayerEvent.java b/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/CancelablePlayerEvent.java
new file mode 100644
index 0000000..4f9b8bc
--- /dev/null
+++ b/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/CancelablePlayerEvent.java
@@ -0,0 +1,42 @@
+package com.vanillage.minecraftalphaserver.event;
+
+import com.vanillage.utils.event.Cancelable;
+import com.vanillage.utils.event.Event;
+import com.vanillage.utils.event.Listener;
+
+import net.minecraft.src.EntityPlayerMP;
+
+public abstract class CancelablePlayerEvent<L extends Listener<L, SELF>, SELF extends Event<L, SELF>> extends PlayerEvent<L, SELF> implements Cancelable {
+    private boolean canceled;
+
+    public CancelablePlayerEvent(EntityPlayerMP player) {
+        this(player, false);
+    }
+
+    public CancelablePlayerEvent(EntityPlayerMP player, boolean canceled) {
+        super(player);
+        this.canceled = canceled;
+    }
+
+    @Override
+    public final boolean isCanceled() {
+        return canceled;
+    }
+
+    @Override
+    public final boolean setCanceled(boolean canceled) {
+        boolean previous = this.canceled;
+        this.canceled = canceled;
+        return previous;
+    }
+
+    @Override
+    public final boolean cancel() {
+        return setCanceled(true);
+    }
+
+    @Override
+    public final boolean uncancel() {
+        return setCanceled(false);
+    }
+}
diff --git a/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/PlayerEvent.java b/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/PlayerEvent.java
new file mode 100644
index 0000000..6477895
--- /dev/null
+++ b/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/PlayerEvent.java
@@ -0,0 +1,22 @@
+package com.vanillage.minecraftalphaserver.event;
+
+import com.vanillage.utils.event.Event;
+import com.vanillage.utils.event.Listener;
+
+import net.minecraft.src.EntityPlayerMP;
+
+public abstract class PlayerEvent<L extends Listener<L, SELF>, SELF extends Event<L, SELF>> extends Event<L, SELF> {
+    private final EntityPlayerMP player;
+
+    public PlayerEvent(EntityPlayerMP player) {
+        if (player == null) {
+            throw new IllegalArgumentException("player cannot be null");
+        }
+
+        this.player = player;
+    }
+
+    public final EntityPlayerMP getPlayer() {
+        return player;
+    }
+}
diff --git a/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/ResultablePlayerEvent.java b/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/ResultablePlayerEvent.java
new file mode 100644
index 0000000..2320728
--- /dev/null
+++ b/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/ResultablePlayerEvent.java
@@ -0,0 +1,121 @@
+package com.vanillage.minecraftalphaserver.event;
+
+import com.vanillage.utils.event.Event;
+import com.vanillage.utils.event.Listener;
+import com.vanillage.utils.event.Resultable;
+
+import net.minecraft.src.EntityPlayerMP;
+
+public abstract class ResultablePlayerEvent<L extends Listener<L, SELF>, SELF extends Event<L, SELF>> extends PlayerEvent<L, SELF> implements Resultable {
+    private Result result;
+
+    public ResultablePlayerEvent(EntityPlayerMP player) {
+        this(player, Result.HANDLE);
+    }
+
+    public ResultablePlayerEvent(EntityPlayerMP player, Result result) {
+        super(player);
+
+        if (result == null) {
+            throw new IllegalArgumentException("result cannot be null");
+        }
+
+        this.result = result;
+    }
+
+    @Override
+    public final Result getResult() {
+        return result;
+    }
+
+    @Override
+    public final Result setResult(Result result) {
+        if (result == null) {
+            throw new IllegalArgumentException("result cannot be null");
+        }
+
+        Result previous = this.result;
+        this.result = result;
+        return previous;
+    }
+
+    @Override
+    public final boolean isHandled() {
+        return result == Result.HANDLE;
+    }
+
+    @Override
+    public final boolean isAllowed() {
+        return result == Result.ALLOW;
+    }
+
+    @Override
+    public final boolean isDenied() {
+        return result == Result.DENY;
+    }
+
+    @Override
+    public final Result handle() {
+        return setResult(Result.HANDLE);
+    }
+
+    @Override
+    public final Result allow() {
+        return setResult(Result.ALLOW);
+    }
+
+    @Override
+    public final Result deny() {
+        return setResult(Result.DENY);
+    }
+
+    @Override
+    public final Result setResultFromTo(Result from, Result to) {
+        return result == from ? setResult(to) : result;
+    }
+
+    @Override
+    public final Result handleIf(Result result) {
+        return setResultFromTo(result, Result.HANDLE);
+    }
+
+    @Override
+    public final Result allowIf(Result result) {
+        return setResultFromTo(result, Result.ALLOW);
+    }
+
+    @Override
+    public final Result denyIf(Result result) {
+        return setResultFromTo(result, Result.DENY);
+    }
+
+    @Override
+    public final Result handleIfAllowed() {
+        return handleIf(Result.ALLOW);
+    }
+
+    @Override
+    public final Result handleIfDenied() {
+        return handleIf(Result.DENY);
+    }
+
+    @Override
+    public final Result allowIfHandled() {
+        return allowIf(Result.HANDLE);
+    }
+
+    @Override
+    public final Result allowIfDenied() {
+        return allowIf(Result.DENY);
+    }
+
+    @Override
+    public final Result denyIfHandled() {
+        return denyIf(Result.HANDLE);
+    }
+
+    @Override
+    public final Result denyIfAllowed() {
+        return denyIf(Result.ALLOW);
+    }
+}
diff --git a/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/TestEvent.java b/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/TestEvent.java
new file mode 100644
index 0000000..535c907
--- /dev/null
+++ b/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/TestEvent.java
@@ -0,0 +1,26 @@
+package com.vanillage.minecraftalphaserver.event;
+
+import com.vanillage.utils.event.Event;
+
+import net.minecraft.server.MinecraftServer;
+
+public final class TestEvent extends Event<TestListener, TestEvent> {
+    private final MinecraftServer minecraftServer;
+
+    public TestEvent(MinecraftServer minecraftServer) {
+        if (minecraftServer == null) {
+            throw new IllegalArgumentException("minecraftServer cannot be null");
+        }
+
+        this.minecraftServer = minecraftServer;
+    }
+
+    public MinecraftServer getMinecraftServer() {
+        return minecraftServer;
+    }
+
+    @Override
+    protected Class<TestListener> getListenerClass() {
+        return TestListener.class;
+    }
+}
diff --git a/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/TestListener.java b/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/TestListener.java
new file mode 100644
index 0000000..62d17f5
--- /dev/null
+++ b/sources/minecraft_server/com/vanillage/minecraftalphaserver/event/TestListener.java
@@ -0,0 +1,18 @@
+package com.vanillage.minecraftalphaserver.event;
+
+import com.vanillage.utils.event.Listener;
+
+public abstract class TestListener extends Listener<TestListener, TestEvent> {
+    public TestListener() {
+        this(0);
+    }
+
+    public TestListener(int priority) {
+        super(priority);
+    }
+
+    @Override
+    protected final Class<TestListener> getListenerClass() {
+        return TestListener.class;
+    }
+}
diff --git a/sources/minecraft_server/com/vanillage/minecraftalphaserver/plugin/PluginManager.java b/sources/minecraft_server/com/vanillage/minecraftalphaserver/plugin/PluginManager.java
index 37eb483..cddb7df 100644
--- a/sources/minecraft_server/com/vanillage/minecraftalphaserver/plugin/PluginManager.java
+++ b/sources/minecraft_server/com/vanillage/minecraftalphaserver/plugin/PluginManager.java
@@ -1,5 +1,6 @@
 package com.vanillage.minecraftalphaserver.plugin;
 
+import com.vanillage.utils.event.EventManager;
 import com.vanillage.utils.plugin.SimplePluginLoader;
 import com.vanillage.utils.plugin.SimplePluginManager;
 
@@ -7,11 +8,16 @@ import net.minecraft.server.MinecraftServer;
 
 public final class PluginManager extends SimplePluginManager<Plugin> {
     private final MinecraftServer minecraftServer;
+    private final EventManager eventManager = new EventManager();
 
     public PluginManager(MinecraftServer minecraftServer) {
         this.minecraftServer = minecraftServer;
     }
 
+    public EventManager getEventManager() {
+        return eventManager;
+    }
+
     @Override
     protected void enablePlugin(Plugin plugin) {
         plugin.setMinecraftServer(minecraftServer);
diff --git a/sources/minecraft_server/com/vanillage/utils/ClassToInstanceListHashMap.java b/sources/minecraft_server/com/vanillage/utils/ClassToInstanceListHashMap.java
new file mode 100644
index 0000000..20917dd
--- /dev/null
+++ b/sources/minecraft_server/com/vanillage/utils/ClassToInstanceListHashMap.java
@@ -0,0 +1,19 @@
+package com.vanillage.utils;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ClassToInstanceListHashMap {
+    private final Map<Class<?>, List<?>> map = new HashMap<>();
+
+    @SuppressWarnings("unchecked")
+    public <T> List<T> put(Class<T> key, List<T> value) {
+        return (List<T>) map.put(key, value);
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T> List<T> get(Class<T> key) {
+        return (List<T>) map.get(key);
+    }
+}
diff --git a/sources/minecraft_server/com/vanillage/utils/event/Cancelable.java b/sources/minecraft_server/com/vanillage/utils/event/Cancelable.java
new file mode 100644
index 0000000..33bd3b0
--- /dev/null
+++ b/sources/minecraft_server/com/vanillage/utils/event/Cancelable.java
@@ -0,0 +1,11 @@
+package com.vanillage.utils.event;
+
+public interface Cancelable {
+    boolean isCanceled();
+
+    boolean setCanceled(boolean canceled);
+
+    boolean cancel();
+
+    boolean uncancel();
+}
diff --git a/sources/minecraft_server/com/vanillage/utils/event/CancelableEvent.java b/sources/minecraft_server/com/vanillage/utils/event/CancelableEvent.java
new file mode 100644
index 0000000..15f55c7
--- /dev/null
+++ b/sources/minecraft_server/com/vanillage/utils/event/CancelableEvent.java
@@ -0,0 +1,35 @@
+package com.vanillage.utils.event;
+
+public abstract class CancelableEvent<L extends Listener<L, SELF>, SELF extends Event<L, SELF>> extends Event<L, SELF> implements Cancelable {
+    private boolean canceled;
+
+    public CancelableEvent() {
+        this(false);
+    }
+
+    public CancelableEvent(boolean canceled) {
+        this.canceled = canceled;
+    }
+
+    @Override
+    public final boolean isCanceled() {
+        return canceled;
+    }
+
+    @Override
+    public final boolean setCanceled(boolean canceled) {
+        boolean previous = this.canceled;
+        this.canceled = canceled;
+        return previous;
+    }
+
+    @Override
+    public final boolean cancel() {
+        return setCanceled(true);
+    }
+
+    @Override
+    public final boolean uncancel() {
+        return setCanceled(false);
+    }
+}
diff --git a/sources/minecraft_server/com/vanillage/utils/event/Event.java b/sources/minecraft_server/com/vanillage/utils/event/Event.java
new file mode 100644
index 0000000..8809ec2
--- /dev/null
+++ b/sources/minecraft_server/com/vanillage/utils/event/Event.java
@@ -0,0 +1,5 @@
+package com.vanillage.utils.event;
+
+public abstract class Event<L extends Listener<L, SELF>, SELF extends Event<L, SELF>> {
+    protected abstract Class<L> getListenerClass();
+}
diff --git a/sources/minecraft_server/com/vanillage/utils/event/EventManager.java b/sources/minecraft_server/com/vanillage/utils/event/EventManager.java
new file mode 100644
index 0000000..25716b3
--- /dev/null
+++ b/sources/minecraft_server/com/vanillage/utils/event/EventManager.java
@@ -0,0 +1,50 @@
+package com.vanillage.utils.event;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+
+import com.vanillage.utils.ClassToInstanceListHashMap;
+
+public class EventManager {
+    private final ClassToInstanceListHashMap listeners = new ClassToInstanceListHashMap();
+
+    public <L extends Listener<L, E>, E extends Event<L, E>> void registerListener(L listener) {
+        List<L> listeners = this.listeners.get(listener.getListenerClass());
+
+        if (listeners == null) {
+            listeners = new LinkedList<>();
+            this.listeners.put(listener.getListenerClass(), listeners);
+        }
+
+        ListIterator<L> iterator = listeners.listIterator();
+        boolean higherPriority = true;
+
+        while (iterator.hasNext() && (higherPriority = iterator.next().getPriority() <= listener.getPriority())) {
+
+        }
+
+        if (!higherPriority) {
+            iterator.previous();
+        }
+
+        iterator.add(listener);
+    }
+
+    public <L extends Listener<L, E>, E extends Event<L, E>> void callEvent(E event) {
+        List<L> listeners = this.listeners.get(event.getListenerClass());
+
+        if (listeners == null) {
+            return;
+        }
+
+        for (L listener : listeners) {
+            try {
+                listener.onEvent(event);
+            } catch (Exception e) {
+                System.err.println("error while calling event");
+                e.printStackTrace();
+            }
+        }
+    }
+}
diff --git a/sources/minecraft_server/com/vanillage/utils/event/Listener.java b/sources/minecraft_server/com/vanillage/utils/event/Listener.java
new file mode 100644
index 0000000..7e8c6c0
--- /dev/null
+++ b/sources/minecraft_server/com/vanillage/utils/event/Listener.java
@@ -0,0 +1,23 @@
+package com.vanillage.utils.event;
+
+public abstract class Listener<SELF extends Listener<SELF, E>, E extends Event<SELF, E>> {
+    private final int priority;
+
+    // Force direct subclass of this class to define a constructor
+    // All direct subclasses of this class should care about the priority
+    /*public Listener() {
+        this(0);
+    }*/
+
+    public Listener(int priority) {
+        this.priority = priority;
+    }
+
+    protected abstract void onEvent(E event);
+
+    protected abstract Class<SELF> getListenerClass();
+
+    protected final int getPriority() {
+        return priority;
+    }
+}
diff --git a/sources/minecraft_server/com/vanillage/utils/event/Resultable.java b/sources/minecraft_server/com/vanillage/utils/event/Resultable.java
new file mode 100644
index 0000000..5c03142
--- /dev/null
+++ b/sources/minecraft_server/com/vanillage/utils/event/Resultable.java
@@ -0,0 +1,43 @@
+package com.vanillage.utils.event;
+
+public interface Resultable {
+    Result getResult();
+
+    Result setResult(Result result);
+
+    boolean isHandled();
+
+    boolean isAllowed();
+
+    boolean isDenied();
+
+    Result handle();
+
+    Result allow();
+
+    Result deny();
+
+    Result setResultFromTo(Result from, Result to);
+
+    Result handleIf(Result result);
+
+    Result allowIf(Result result);
+
+    Result denyIf(Result result);
+
+    Result handleIfAllowed();
+
+    Result handleIfDenied();
+
+    Result allowIfHandled();
+
+    Result allowIfDenied();
+
+    Result denyIfHandled();
+
+    Result denyIfAllowed();
+
+    public enum Result {
+        HANDLE, ALLOW, DENY
+    }
+}
diff --git a/sources/minecraft_server/com/vanillage/utils/event/ResultableEvent.java b/sources/minecraft_server/com/vanillage/utils/event/ResultableEvent.java
new file mode 100644
index 0000000..e49b51c
--- /dev/null
+++ b/sources/minecraft_server/com/vanillage/utils/event/ResultableEvent.java
@@ -0,0 +1,113 @@
+package com.vanillage.utils.event;
+
+public abstract class ResultableEvent<L extends Listener<L, SELF>, SELF extends Event<L, SELF>> extends Event<L, SELF> implements Resultable {
+    private Result result;
+
+    public ResultableEvent() {
+        this(Result.HANDLE);
+    }
+
+    public ResultableEvent(Result result) {
+        if (result == null) {
+            throw new IllegalArgumentException("result cannot be null");
+        }
+
+        this.result = result;
+    }
+
+    @Override
+    public final Result getResult() {
+        return result;
+    }
+
+    @Override
+    public final Result setResult(Result result) {
+        if (result == null) {
+            throw new IllegalArgumentException("result cannot be null");
+        }
+
+        Result previous = this.result;
+        this.result = result;
+        return previous;
+    }
+
+    @Override
+    public final boolean isHandled() {
+        return result == Result.HANDLE;
+    }
+
+    @Override
+    public final boolean isAllowed() {
+        return result == Result.ALLOW;
+    }
+
+    @Override
+    public final boolean isDenied() {
+        return result == Result.DENY;
+    }
+
+    @Override
+    public final Result handle() {
+        return setResult(Result.HANDLE);
+    }
+
+    @Override
+    public final Result allow() {
+        return setResult(Result.ALLOW);
+    }
+
+    @Override
+    public final Result deny() {
+        return setResult(Result.DENY);
+    }
+
+    @Override
+    public final Result setResultFromTo(Result from, Result to) {
+        return result == from ? setResult(to) : result;
+    }
+
+    @Override
+    public final Result handleIf(Result result) {
+        return setResultFromTo(result, Result.HANDLE);
+    }
+
+    @Override
+    public final Result allowIf(Result result) {
+        return setResultFromTo(result, Result.ALLOW);
+    }
+
+    @Override
+    public final Result denyIf(Result result) {
+        return setResultFromTo(result, Result.DENY);
+    }
+
+    @Override
+    public final Result handleIfAllowed() {
+        return handleIf(Result.ALLOW);
+    }
+
+    @Override
+    public final Result handleIfDenied() {
+        return handleIf(Result.DENY);
+    }
+
+    @Override
+    public final Result allowIfHandled() {
+        return allowIf(Result.HANDLE);
+    }
+
+    @Override
+    public final Result allowIfDenied() {
+        return allowIf(Result.DENY);
+    }
+
+    @Override
+    public final Result denyIfHandled() {
+        return denyIf(Result.HANDLE);
+    }
+
+    @Override
+    public final Result denyIfAllowed() {
+        return denyIf(Result.ALLOW);
+    }
+}
diff --git a/sources/minecraft_server/net/minecraft/server/MinecraftServer.java b/sources/minecraft_server/net/minecraft/server/MinecraftServer.java
index 69d0a7f..3313310 100644
--- a/sources/minecraft_server/net/minecraft/server/MinecraftServer.java
+++ b/sources/minecraft_server/net/minecraft/server/MinecraftServer.java
@@ -4,6 +4,7 @@
 
 package net.minecraft.server;
 
+import com.vanillage.minecraftalphaserver.event.TestEvent;//Mod: Add basic event system
 import com.vanillage.minecraftalphaserver.plugin.Plugin;//Mod: Add basic plugin system
 import com.vanillage.minecraftalphaserver.plugin.PluginManager;//Mod: Add basic plugin system
 import com.vanillage.minecraftalphaserver.plugin.PluginManagerProvider;//Mod: Add basic plugin system
@@ -111,6 +112,7 @@ public class MinecraftServer
         pluginManagerProvider = new PluginManagerProvider(this);//Mod: Add basic plugin system
         pluginManager = pluginManagerProvider.getPluginManager();//Mod: Add basic plugin system
         pluginManagerProvider.loadPlugins(new JarsDirectoryPluginLoader<>(Plugin.class));//Mod: Add basic plugin system
+        pluginManager.getEventManager().callEvent(new TestEvent(this));//Mod: Add basic event system
         logger.info("Done! For help, type \"help\" or \"?\"");
         return true;
     }
